plugins/ccgx/fu-ccgx-dock-bb.c

```diff
@@ -0,0 +1,734 @@
+	return TRUE;
+}
+
+/* write firmware for pd i2c device */
+static gboolean
+fu_ccgx_dock_bb_pd_i2c_write_fw (FuDevice *device,
+				 guint8 *fw_buffer,
+				 gsize fw_size,
+				 FwupdInstallFlags flags,
+				 GError **error)
+{
+	FuCcgxDockBb *self = FU_CCGX_DOCK_BB (device);
+	CyHPIHandle *hpi_handle = self->hpi_handle;
+	guint32 fw_pos = 0;
+	guint16 row_num = 0;
+	guint16 row_size = 0;
+	guint8 *row_data = NULL;
+	FWMode update_fw_mode = 0;
+	CyacdFileInfo cyacd_file_info = {0};
+	CyacdFileHandle cyacd_handle_array[CYACD_HANDLE_MAX_COUNT] = {0};
+	CyacdFileHandle* cyacd_handle = NULL;
+	guint32 handle_count = 0;
+	guint32 index = 0;
+	CCGxMetaData* metadata = NULL;
+	PDFWAppVersion update_fw_version = {0};
+	gsize update_fw_size = 0;
+	g_autofree gchar *update_str_version = NULL;
+	g_autofree guint8 *row_buffer = g_malloc0 (CYACD_ROW_BUFFER_SIZE);
+
+	handle_count =  fu_ccgx_cyacd_file_init_handle (cyacd_handle_array,
+							CYACD_HANDLE_MAX_COUNT,
+							fw_buffer, fw_size);
+
+	if (self->fw_image_type != FW_IMAGE_TYPE_DUAL_SYMMETRIC &&
+		 self->fw_image_type != FW_IMAGE_TYPE_DUAL_ASYMMETRIC) {
+			g_set_error(error,FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				   "not supported fw image");
+		return FALSE;
+	}
+
+	if (handle_count <= 0) {
+		g_set_error(error,FWUPD_ERROR,
+			   FWUPD_ERROR_NOT_SUPPORTED,
+			   "invalid firmware type");
+		return FALSE;
+	}
+
+	for (index = 0; index < handle_count ; index++) {
+		/* get cyacd handle */
+		cyacd_handle = &cyacd_handle_array [index];
+		update_fw_size = cyacd_handle->buffer_size;
+
+		/* parse cyacd data */
+		if (!fu_ccgx_cyacd_file_parse (cyacd_handle, &cyacd_file_info)) {
+			g_set_error (error,FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "cyacd parsing error");
+			g_warning ("cyacd parsing error");
+			return FALSE;
+		}
+
+		/* check silicon ID */
+		if (self->pd_device_data.silicon_id != cyacd_file_info.silicon_id) {
+			g_set_error (error,FWUPD_ERROR,
+				     FWUPD_ERROR_NOT_SUPPORTED,
+				     "silicon id mismatch");
+			g_warning ("silicon id mismatch 0x%X / 0x%X ", self->pd_device_data.silicon_id,
+				   cyacd_file_info.silicon_id);
+			return FALSE;
+		}
+
+		/* check application version type */
+		if (self->pd_device_data.fw_mode != FW_MODE_BOOT) {
+			if (self->pd_device_data.current_version.ver.type != cyacd_file_info.app_version.ver.type) {
+				g_set_error (error, FWUPD_ERROR,
+					     FWUPD_ERROR_NOT_SUPPORTED,
+					     "application type mismatch");
+				g_warning ("application	type mismatch 0x%02X / 0x%02X", 
+					   self->pd_device_data.current_version.ver.type ,
+					   cyacd_file_info.app_version.ver.type);
+				return FALSE;
+			}
+		}
+
+		if (self->fw_image_type == FW_IMAGE_TYPE_DUAL_SYMMETRIC || self->fw_image_type == FW_IMAGE_TYPE_DUAL_ASYMMETRIC) {
+			if (self->pd_device_data.fw_mode == FW_MODE_FW1)
+				update_fw_mode = FW_MODE_FW2;
+			else if (self->pd_device_data.fw_mode == FW_MODE_FW2)
+				update_fw_mode = FW_MODE_FW1;
```

Comment:
just to clarify, it looks like we do two firmware updates, spaced apart. Why do we need to upgrade the backup FW1 at all? Can't we keep that on factory-set default and only update the FW2 primary image?

Reply:
Cypress Dock supports Asymmetric Firmware and Symmetric Firmware design.
Current target is gen2 dock and gen2 support Asymmetric Firmware .
hybrid dock is next target and support symmetric Firmware .

If FW1 is running, FW2 can be updated in FW1
If FW2 is running, FW1 can be updated in FW2

In Asymmetric Firmware, FW1 is Backup image and FW2 is Primary image.
In Symmetric Firmware, both FW1 and FW2 is Primary image.

Primary FW image support full functions of Dock while backup image support limited functions: e.g. PD charging support but no Display port or TBT.

For Asymmetric Firmware.
 - Device is normally run on FW2 (Primary).
 - In FW Update, if device is running on FW2 (Primary), we need two steps.
   1) plugin detect FW2-> notify fwupd of fw2 version -> plugin update FW1 (Backup) -> reboot 
   2) plugin detect FW1 -> notify fwupd of 0.0.0 version to make sure FW2 update -> plugin update FW2 (Primary) -> reboot.
 - In FW Update, if device is running on FW1 (Backup), we need one steps.
   1) plugin detect FW1 -> notify fwupd of 0.0.0 version to make sure FW2 update -> plugin update FW2 (Primary) -> reboot.

For symmetric Firmware.
  - Device can run on FW1 and FW2.
  - In FW Update, we need one steps. 
   1) plugin detect FW1 or FW2 -> notify fwupd of FW1 or FW2 version  -> plugin update FW1 or FW2 -> reboot

Reply:
How do we tell symmetric devices which firmware to boot? It seems complicated to update both when we could just update FW2 and leave FW1 as a known-good factory-supplied version.

Reply:
I add additional information.
If FW1 is running, FW2 will be updated, after FW2 update and reboot, FW2 is running vice versa.

Reply:
For dock kind of application, symmetric fw (both FW1 and FW2 has same size and same functionality) actually helps to keeps the user experience unified across both Fw images. There were few initial dock solutions which used asymmetric fw image. 
Asymmetric FW has a backup fw and a main or primary fw. 
Back up fw has a smaller size fw image with minimum functionality and mostly used to update main fw.

Each FW image has a meta data associated with it. 

For symmetric Fw images: device fw checks fw metadata to decide which fw image is latest and starts running from latest FW image. So device boots from alternate fw post power cycle. So we can update only one image as Ryan has mentioned above.

For asymmetric Fw images: We need to update both images. Device always runs from Primary Fw. Typical FW update sequence:
1. Update back fw.
2. Issue jump to alt fw image request.
3. Device reboots from backup fw.
4. Update primary Fw.
5. Issue jump to alt fw image request.
6. Device reboots from primary fw.

Yes we can update only main or primary image. But this would not not help in user experience. In either case, we need to switch from current or primary fw to backup fw and then update primary fw or main fw. And again issue a reset to switch to primary Fw. So we just saved writing backup fw in 1st step when primary fw was running. This anyhow user would not have seen it would have happened in background.

Reply:
Symmetric fw update requires: Both images of same size and hence flash size available on device comes into picture here. 
If device does not have sufficient flash we go with asymmetric fw image. Primary fw has all the functionality and backup fw has limited or minimum functionality. Typically it supports USB Type-C PD functionality to power up the devices connected through it and allow fw update.